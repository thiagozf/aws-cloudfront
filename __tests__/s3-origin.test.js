const {
  mockCreateDistribution,
  mockUpdateDistribution,
  mockCreateDistributionPromise,
  mockGetDistributionConfigPromise,
  mockUpdateDistributionPromise,
  mockCreateCloudFrontOriginAccessIdentityPromise
} = require('aws-sdk')

const { createComponent } = require('../test-utils')

describe('Input origin as an S3 bucket url', () => {
  let component

  beforeEach(async () => {
    mockCreateDistributionPromise.mockResolvedValueOnce({
      Distribution: {
        Id: 'distributionwithS3origin'
      }
    })

    component = await createComponent()
  })

  it('creates distribution with S3 origin', async () => {
    await component.default({
      origins: ['https://mybucket.s3.amazonaws.com']
    })

    expect(mockCreateDistribution).toBeCalledWith(
      expect.objectContaining({
        DistributionConfig: expect.objectContaining({
          Origins: expect.objectContaining({
            Items: [
              {
                Id: 'mybucket',
                DomainName: 'mybucket.s3.amazonaws.com',
                S3OriginConfig: {
                  OriginAccessIdentity: ''
                },
                CustomHeaders: {
                  Quantity: 0,
                  Items: []
                },
                OriginPath: ''
              }
            ]
          })
        })
      })
    )
    expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot()
  })

  it('creates distribution configured to serve private S3 content', async () => {
    mockCreateCloudFrontOriginAccessIdentityPromise.mockResolvedValueOnce({
      CloudFrontOriginAccessIdentity: {
        Id: 'access-identity-xyz',
        S3CanonicalUserId: 's3-canonical-user-id-xyz'
      }
    })

    await component.default({
      origins: [
        {
          url: 'https://mybucket.s3.amazonaws.com',
          private: true
        }
      ]
    })

    expect(mockCreateDistribution).toBeCalledWith(
      expect.objectContaining({
        DistributionConfig: expect.objectContaining({
          Origins: expect.objectContaining({
            Items: [
              {
                Id: 'mybucket',
                DomainName: 'mybucket.s3.amazonaws.com',
                S3OriginConfig: {
                  OriginAccessIdentity: 'origin-access-identity/cloudfront/access-identity-xyz'
                },
                CustomHeaders: {
                  Quantity: 0,
                  Items: []
                },
                OriginPath: ''
              }
            ]
          })
        })
      })
    )
    expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot()
  })

  it('updates distribution configured to serve private S3 content', async () => {
    mockCreateCloudFrontOriginAccessIdentityPromise.mockResolvedValue({
      CloudFrontOriginAccessIdentity: {
        Id: 'access-identity-xyz',
        S3CanonicalUserId: 's3-canonical-user-id-xyz'
      }
    })

    mockGetDistributionConfigPromise.mockResolvedValueOnce({
      ETag: 'etag',
      DistributionConfig: {
        Origins: {
          Items: []
        }
      }
    })

    mockUpdateDistributionPromise.mockResolvedValueOnce({
      Distribution: {
        Id: 'distributionwithS3originupdated'
      }
    })

    await component.default({
      origins: [
        {
          url: 'https://mybucket.s3.amazonaws.com',
          private: true
        }
      ]
    })

    await component.default({
      origins: [
        {
          url: 'https://anotherbucket.s3.amazonaws.com',
          private: true
        }
      ]
    })

    expect(mockUpdateDistribution).toBeCalledWith(
      expect.objectContaining({
        DistributionConfig: expect.objectContaining({
          Origins: expect.objectContaining({
            Items: [
              {
                Id: 'anotherbucket',
                DomainName: 'anotherbucket.s3.amazonaws.com',
                S3OriginConfig: {
                  OriginAccessIdentity: 'origin-access-identity/cloudfront/access-identity-xyz'
                },
                OriginPath: '',
                CustomHeaders: { Items: [], Quantity: 0 }
              }
            ]
          })
        })
      })
    )
    expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot()
  })

  it('updates distribution', async () => {
    mockGetDistributionConfigPromise.mockResolvedValueOnce({
      ETag: 'etag',
      DistributionConfig: {
        Origins: {
          Items: []
        }
      }
    })
    mockUpdateDistributionPromise.mockResolvedValueOnce({
      Distribution: {
        Id: 'distributionwithS3originupdated'
      }
    })

    await component.default({
      origins: ['https://mybucket.s3.amazonaws.com']
    })

    await component.default({
      origins: ['https://anotherbucket.s3.amazonaws.com']
    })

    expect(mockUpdateDistribution).toBeCalledWith(
      expect.objectContaining({
        DistributionConfig: expect.objectContaining({
          Origins: expect.objectContaining({
            Items: [
              expect.objectContaining({
                Id: 'anotherbucket',
                DomainName: 'anotherbucket.s3.amazonaws.com'
              })
            ]
          })
        })
      })
    )

    expect(mockUpdateDistribution.mock.calls[0][0]).toMatchSnapshot()
  })
})
