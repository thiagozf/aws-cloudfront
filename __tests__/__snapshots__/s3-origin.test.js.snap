// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input origin as an S3 bucket url creates distribution configured to serve private S3 content 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "mybucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "mybucket.s3.amazonaws.com",
          "Id": "mybucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "s3-canonical-user-id-xyz",
          },
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`Input origin as an S3 bucket url creates distribution with S3 origin 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "mybucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "mybucket.s3.amazonaws.com",
          "Id": "mybucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "",
          },
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`Input origin as an S3 bucket url updates distribution 1`] = `
Object {
  "DistributionConfig": Object {
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "anotherbucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "anotherbucket.s3.amazonaws.com",
          "Id": "anotherbucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "",
          },
        },
      ],
      "Quantity": 1,
    },
  },
  "Id": "distributionwithS3origin",
  "IfMatch": "etag",
}
`;
